fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/MembershipIndexManager.java

```diff
@@ -0,0 +1,630 @@
+            DbPlatform.MARIADB, "sql/mariadb-membership.sql"
+    );
+
+    @PostConstruct
+    public void setUp() {
+        jdbcTemplate = new NamedParameterJdbcTemplate(getDataSource());
+
+        final var dbPlatform = DbPlatform.fromDataSource(dataSource);
+
+        Preconditions.checkArgument(DDL_MAP.containsKey(dbPlatform),
+                "Missing DDL mapping for %s", dbPlatform);
+
+        final var ddl = DDL_MAP.get(dbPlatform);
+        log.debug("Applying ddl: {}", ddl);
+        DatabasePopulatorUtils.execute(
+                new ResourceDatabasePopulator(new DefaultResourceLoader().getResource("classpath:" + ddl)),
+                dataSource);
+    }
+
+    /**
+     * End a membership entry, setting an end time if committed, or clearing from the current tx
+     * if it was newly added.
+     *
+     * @param txId transaction id
+     * @param sourceId ID of the direct/indirect container whose membership should be ended
+     * @param membership membership triple to end
+     * @param endTime the time the resource was deleted, generally its last modified
+     */
+    @Transactional
+    public void endMembership(final String txId,  final FedoraId sourceId, final Triple membership,
+            final Instant endTime) {
+        final Map<String, Object> parameterSource = Map.of(
+                TX_ID_PARAM, txId,
+                SOURCE_ID_PARAM, sourceId.getFullId(),
+                SUBJECT_ID_PARAM, membership.getSubject().getURI(),
+                PROPERTY_PARAM, membership.getPredicate().getURI(),
+                OBJECT_ID_PARAM, membership.getObject().getURI(),
+                OPERATION_PARAM, ADD_OPERATION);
+
+        final int affected = jdbcTemplate.update(CLEAR_ENTRY_IN_TX, parameterSource);
+
+        // If no rows were deleted, then assume we need to delete permanent entry
+        if (affected == 0) {
+            final Map<String, Object> parameterSource2 = Map.of(
+                    TX_ID_PARAM, txId,
+                    SOURCE_ID_PARAM, sourceId.getFullId(),
+                    SUBJECT_ID_PARAM, membership.getSubject().getURI(),
+                    PROPERTY_PARAM, membership.getPredicate().getURI(),
+                    OBJECT_ID_PARAM, membership.getObject().getURI(),
+                    END_TIME_PARAM, Timestamp.from(endTime),
+                    NO_END_TIME_PARAM, NO_END_TIMESTAMP,
+                    DELETE_OP_PARAM, DELETE_OPERATION);
+            jdbcTemplate.update(END_EXISTING_MEMBERSHIP, parameterSource2);
+        }
+    }
+
+    /**
+     * End membership entry within a transaction
+     * @param txId transaction id
+     * @param sourceId ID of the direct/indirect container whose membership should be ended
+     * @param membership membership triple to end
+     * @param endTime the time the resource was deleted, generally its last modified
+     */
+    public void endMembershipInTx(final String txId,  final FedoraId sourceId, final Triple membership,
```

Comment:
Can this method be removed? It does not appear to be called anywhere.

Reply:
Good catch, I think this got leftover during a few refactors of `populateMembershipHistory`